composer/algorithms/gated_linear_units/gated_linear_units.py

```diff
@@ -0,0 +1,165 @@
+    return IdentityLayer()
+
+
+def apply_gated_linear_units(model: torch.nn.Module,
+                             optimizers: Union[torch.optim.Optimizer, Sequence[torch.optim.Optimizer]],
+                             act_fn: Optional[Callable] = None,
+                             gated_layer_bias: bool = False,
+                             non_gated_layer_bias: bool = False) -> None:
+    """
+    Replaces the Linear layers in the feed-forward network with `Gated Linear Units <https://arxiv.org/abs/2002.05202>`_.
+
+    Args:
+        model (:class:`torch.nn.Module`): the model to modify in-place
+        optimizers (:class:`torch.optim.Optimizer` | Sequence[:class:`torch.optim.Optimizer`], optional):
+            Existing optimizers bound to ``model.parameters()``. All optimizers that have already been
+            constructed with ``model.parameters()`` must be specified here so that
+            they will optimize the correct parameters.
+
+            If the optimizer(s) are constructed *after* calling this function,
+            then it is safe to omit this parameter. These optimizers will see the correct
+            model parameters.
+        act_fn (Callable, optional): Optionally, the activation function to use. If ``None``, the algorithm will
+            use the existing activation function in the model.
+        gated_layer_bias (bool, optional): Whether to use biases in the linear layers within the GLU. Default: ``False``.
+        non_gated_layer_bias (bool, optional): Whether to use biases in the linear layers within the GLU. Default: ``False``.
+    """
+    raise_if_missing_transformers(TRANSFORMERS_INSTALLED)
+
+    # ensure that the model is an instance of a BERTModel, since our replacement policy is only defined for BERTs
+    if not isinstance(model, BERTModel):
+        raise ValueError('Gated Linear Units only has a surgery policy defined for instances of BERTModel.')
+
+    if act_fn is None:
+        # get the activation functions used
+        act_fns = {module.intermediate_act_fn for module in model.modules() if isinstance(module, BertIntermediate)}
+
+        if len(act_fns) != 1:
+            raise ValueError('The model has non-uniform activation functions, which is currently unsupported.')
+
+        # since our set is of length-1, let's extract the only activation function remaining.
+        (act_fn,) = act_fns
+
+    if act_fn is None:
+        raise ValueError(
+            'Could not find an existing activation function to use, and no custom activation function was provided.')
+
+    # prepare the replacement policy and perform replacement
+    policy: Dict[Type[torch.nn.Module], module_surgery.ReplacementFunction] = {
+        BertIntermediate:
+            from_BertIntermediate,
+        BertOutput:
+            partial(from_BertOutput,
```

Comment:
nit: For typing, would it be possible to define a sub-function here to bind these parameters, rather than use `functools.partial`? I don't believe pyright can validate functools.partial argument names or types...(I might be mistaken given the new ParamSpec in python 3.10, in which case ignore this comment)